### PR TITLE
Speedup function analysis

### PR DESCRIPTION
**Pull request by SchumBlubBlub**
https://github.com/radare/radare2/issues/5274#issuecomment-230157691

Started:
r2 -S ../bins/torivahti.exe -qc 'aac;Ps b'  197.53s user 0.16s system 98% cpu 3:31.82 total

Then:
<del>r2 -S ../bins/torivahti.exe -qc 'aac;Ps b'  56.20s user 0.13s system 98% cpu 57.690 total</del>
<del>r2 -S ../bins/torivahti.exe -qc 'aac;Ps b'  31.13s user 0.13s system 98% cpu 31.690 total</del>
<del>r2 -S ../bins/torivahti.exe -qc 'aac;Ps b'  25.29s user 0.13s system 98% cpu 25.758 total</del>
r2 -S ../bins/torivahti.exe -qc 'aac;Ps b'  22.97s user 0.12s system 98% cpu 23.362 total

Check results:
```
$ r2-master -S ../bins/torivahti.exe -qc 'aac;Ps a'                            
r2 -S ../bins/torivahti.exe -qc 'aac;Ps b'  197.53s user 0.16s system 98% cpu 3:31.82 total
$ r2 -S ../bins/torivahti.exe -qc 'aac;Ps b'  
r2 -S ../bins/torivahti.exe -qc 'aac;Ps a'  46.36s user 0.36s system 22% cpu 3:30.21 total
$  diff ~/.config/radare2/projects/b/rc ~/.config/radare2/projects/a/rc
< "e cfg.user = pid12045"
---
> "e cfg.user = pid843"
103561,103562c103561,103562
< f sym.___chkstk 50 0x00644a50
< af+ 0x00644a50 50 sym.___chkstk f n
---
> f sym.__alloca 50 0x00644a50
> af+ 0x00644a50 50 sym.__alloca f n

$ grep "^f" ~/.config/radare2/projects/b/rc | wc -l
29989

```

Todo:

- [ ] Fix r2r Regressions
- [x] Test if  flag_get_at can use get_i when `closest=false` instead, so all callsites enjoy (moved to #6757)

### CODE REVIEW

**Comment by condret** [comment]
looks good for me

**Comment by Maijin** [comment]
Doesn't build. https://travis-ci.org/radare/radare2/jobs/199555939#L271

**Comment by SchumBlubBlub** [comment]
funny-golf-animated-gif-6

File: libr/anal/anal.c [hunk 1]
```diff
@@ -564,14 +564,15 @@ R_API bool r_anal_noreturn_at(RAnal *anal, ut64 addr) {
 	int ofs = anal->flb.f->space_idx;
 	anal->flb.set_fs (anal->flb.f, "imports");
 	anal->flb.f->space_strict = true;
-	RFlagItem *fi = anal->flb.get_at (anal->flb.f, addr, false);
+	RFlagItem *fi = r_flag_get_i(anal->flb.f, addr);
```
Comment (radare) [inline libr/anal/anal.c]: missing space before (

File: libr/anal/anal.c [hunk 2]
```diff
@@ -564,14 +564,15 @@ R_API bool r_anal_noreturn_at(RAnal *anal, ut64 addr) {
...
 	if (!fi) {
 		anal->flb.set_fs (anal->flb.f, "symbols");
-		fi = anal->flb.get_at (anal->flb.f, addr, false);
+		fi = r_flag_get_i(anal->flb.f, addr);
 	}
 	anal->flb.f->space_idx = ofs;
 	anal->flb.f->space_strict = oss;
-	if (fi) {
+	if (fi && (!f || strcmp(f->name,fi->name))) {
```
Comment (radare) [inline libr/anal/anal.c]: missing space

File: libr/anal/types.c [hunk 3]
```diff
@@ -369,24 +369,42 @@ R_API char *r_anal_type_func_args_name(RAnal *anal, const char *func_name, int i
...
 
-/* WTF always return 1? */
-static int captureSubString (void *p, const char *k, const char *v) {
-	char **ret = (char **)p;
-	if (strcmp (v, "func")) {
-		return 1;
+#define MIN_MATCH_LEN 4
+
+R_API char *r_anal_type_func_guess(RAnal *anal, char *func_name) {
+	int i = 0, j =0, offset = 0, n;
```
Comment (radare) [inline libr/anal/types.c]: missing space after =

File: libr/anal/types.c [hunk 4]
```diff
@@ -369,24 +369,42 @@ R_API char *r_anal_type_func_args_name(RAnal *anal, const char *func_name, int i
...
+	char* res = NULL;
+
+	size_t slen = strlen(str);
+	if (slen < MIN_MATCH_LEN) {
+		return NULL;
 	}
-	if (strstr (ret[1], k) && (!ret[0] || strlen(k) > strlen (ret[0]))) {
-		free (ret[0]);
-		ret[0] = strdup (k);
+	if((slen > 4 && func_name[4] == '.') &&
```
Comment (radare) [inline libr/anal/types.c]: missing space before (

File: libr/anal/types.c [hunk 5]
```diff
@@ -369,24 +369,42 @@ R_API char *r_anal_type_func_args_name(RAnal *anal, const char *func_name, int i
...
+		((str[0] == 's' && str[1] == 'y'  && str[0] == 'm') ||
+		 (str[0] == 'f' && str[1] == 'u'  && str[0] == 'n')    )) {
+		offset = 4;
+	}
+	str = strdup(&func_name[offset]);
```
Comment (radare) [inline libr/anal/types.c]: missing space before (

File: libr/anal/types.c [hunk 6]
```diff
@@ -369,24 +369,42 @@ R_API char *r_anal_type_func_args_name(RAnal *anal, const char *func_name, int i
...
+	slen -= offset;
+	for (n = slen; n >= MIN_MATCH_LEN; --n) {
+		for (j = 0; j < slen - n ; ++j) {
+			char saved = str[j+n];
+			str[j+n] = 0;
+			if (sdb_exists(anal->sdb_types, &str[j])) {
+				res = sdb_get(anal->sdb_types, str, 0);
+				// XXX if we throw away all of these, why keep them
+				// in same sdb? most types are func. Wasted work.
+				bool is_func = res && !strcmp("func", res);
```
Comment (radare) [inline libr/anal/types.c]: missing space before (

File: libr/anal/types.c [hunk 7]
```diff
@@ -369,24 +369,42 @@ R_API char *r_anal_type_func_args_name(RAnal *anal, const char *func_name, int i
...
+		offset = 4;
+	}
+	str = strdup(&func_name[offset]);
+	slen -= offset;
+	for (n = slen; n >= MIN_MATCH_LEN; --n) {
+		for (j = 0; j < slen - n ; ++j) {
+			char saved = str[j+n];
+			str[j+n] = 0;
+			if (sdb_exists(anal->sdb_types, &str[j])) {
+				res = sdb_get(anal->sdb_types, str, 0);
```
Comment (radare) [inline libr/anal/types.c]: use sdb_const_get() and add space before . otherwise you leak

File: libr/anal/types.c [hunk 8]
```diff
@@ -369,24 +369,42 @@ R_API char *r_anal_type_func_args_name(RAnal *anal, const char *func_name, int i
...
+		 (str[0] == 'f' && str[1] == 'u'  && str[0] == 'n')    )) {
+		offset = 4;
+	}
+	str = strdup(&func_name[offset]);
+	slen -= offset;
+	for (n = slen; n >= MIN_MATCH_LEN; --n) {
+		for (j = 0; j < slen - n ; ++j) {
+			char saved = str[j+n];
+			str[j+n] = 0;
+			if (sdb_exists(anal->sdb_types, &str[j])) {
```
Comment (radare) [inline libr/anal/types.c]: missing space

File: libr/anal/types.c [hunk 9]
```diff
@@ -369,24 +369,42 @@ R_API char *r_anal_type_func_args_name(RAnal *anal, const char *func_name, int i
...
+	if((slen > 4 && func_name[4] == '.') &&
+		((str[0] == 's' && str[1] == 'y'  && str[0] == 'm') ||
+		 (str[0] == 'f' && str[1] == 'u'  && str[0] == 'n')    )) {
+		offset = 4;
+	}
+	str = strdup(&func_name[offset]);
+	slen -= offset;
+	for (n = slen; n >= MIN_MATCH_LEN; --n) {
+		for (j = 0; j < slen - n ; ++j) {
+			char saved = str[j+n];
```
Comment (radare) [inline libr/anal/types.c]: spaces aruond +

File: libr/anal/types.c [hunk 10]
```diff
@@ -369,24 +369,42 @@ R_API char *r_anal_type_func_args_name(RAnal *anal, const char *func_name, int i
...
+			str[j+n] = 0;
+			if (sdb_exists(anal->sdb_types, &str[j])) {
+				res = sdb_get(anal->sdb_types, str, 0);
+				// XXX if we throw away all of these, why keep them
+				// in same sdb? most types are func. Wasted work.
+				bool is_func = res && !strcmp("func", res);
+				if (!is_func) {
+					goto stage_left;
```
Comment (radare) [inline libr/anal/types.c]: just return res, instead of goto?

Comment (radare) [inline libr/anal/types.c]: or return strdup(res) if you use const_get

File: libr/core/canal.c [hunk 11]
```diff
@@ -472,7 +472,7 @@ static int core_anal_fcn(RCore *core, ut64 at, ut64 from, int reftype, int depth
 	}
 	fcn->addr = at;
 	r_anal_fcn_set_size (fcn, 0);
-	RFlagItem *fi = r_flag_get_at (core->flags, at, false);
+	RFlagItem *fi = r_flag_get_i (core->flags, at);
```
Comment (radare) [inline libr/core/canal.c]: at != i. this can produce some regressions

Comment (SchumBlubBlub) [inline libr/core/canal.c]: I thought that's equivalent to get_at(false)?

**Comment by radare** [comment]
get_i = gets the flag at the given address (1 hahstable lookup) (O(1))

get-at = gets closest flag at given address (O(n))

**Comment by radare** [comment]
yes, there are plans to optimize rflags too :P to avoid those O(n) cases and this will heavily increase the performance of r2 . mainly in `pd`

**Comment by SchumBlubBlub** [comment]
Sounds good. Need to get rid of ALL linked lists too.

File: libr/anal/types.c (hunk 10 above)
Comment (SchumBlubBlub) [inline libr/anal/types.c]: I actually prefer everything exit through single chokepoint, more simple to audit. but ok.

**Comment by radare** [comment]
yup

File: libr/core/canal.c (hunk 11 above)
Comment (SchumBlubBlub) [inline libr/core/canal.c]: `R_API RFlagItem *r_flag_get_at(RFlag *f, ut64 off, bool closest)`

**Comment by radare** [comment]
with closest = false, its still iterating over all keys, ignoring the ones that are not in the current space… wondering if we can just if (!closest) return r_flag_get_i() without regressions. can u do a pr with this change to evaluate the change?


R_API RFlagItem *r_flag_get_at(RFlag *f, ut64 off, bool closest) {
        RFlagItem *item, *nice = NULL;
        RListIter *iter;

        r_list_foreach (f->flags, iter, item) {
                if (f->space_strict && IS_IN_SPACE (f, item)) {
                        continue;
                }
                if (item->offset == off) {
                        return evalFlag (f, item);
                }
                if (closest && off > item->offset) {
                        if (!nice || nice->offset < item->offset) {
                                nice = item;
                        }
                }
        }
        return evalFlag (f, nice);
}

File: libr/anal/types.c [hunk 12]
```diff
@@ -372,39 +372,38 @@ R_API char *r_anal_type_func_args_name(RAnal *anal, const char *func_name, int i
...
+			str[j + n] = 0;
+			if (sdb_exists (anal->sdb_types, &str[j])) {
+				res = sdb_const_get (anal->sdb_types, str, 0);
 				// XXX if we throw away all of these, why keep them
 				// in same sdb? most types are func. Wasted work.
-				bool is_func = res && !strcmp("func", res);
+				bool is_func = res && !strcmp ("func", res);
 				if (!is_func) {
-					goto stage_left;
+					return strdup(res);
```
Comment (radare) [inline libr/anal/types.c]: missing space before (.

Comment (SchumBlubBlub) [inline libr/anal/types.c]: *deep breath*

**Comment by SchumBlubBlub** [comment]
Sure... but next hack session.

This breaks some r2r, will see how serious later. hope can be fixed.

**Comment by condret** [comment]
maybe we can do something similar to the pcache storage instead of the lists, or what about trees

File: libr/anal/types.c [hunk 13]
```diff
@@ -369,24 +369,45 @@ R_API char *r_anal_type_func_args_name(RAnal *anal, const char *func_name, int i
...
+		for (j = 0; j < slen - n; ++j) {
+			char saved = str[j + n];
+			str[j + n] = 0;
+			if (sdb_exists (anal->sdb_types, &str[j])) {
+				res = sdb_const_get (anal->sdb_types, str, 0);
+				// XXX if we throw away all of these, why keep them
+				// in same sdb? most types are func. Wasted work.
+				bool is_func = res && !strcmp ("func", res);
+				if (!is_func) {
+					return strdup(str);
```
Comment (radare) [inline libr/anal/types.c]: missing space before (

**Comment by radare** [comment]
There are several regressions here, mainly for function signatures , see travis output

**Comment by Maijin** [comment]
It also doesn't build on OSX https://travis-ci.org/radare/radare2/jobs/199676724#L150

**Comment by SchumBlubBlub** [comment]
@radare, you asked me to change this to that `r_flag_get_at` uses `r_flag_get_i` for `!closest`, which is much better then changing the calls directly. But get_flag_i depends on the r_hashtable64, which is buggy so this cause regressions, and you didn't want fixes for that. So I split such part (most of the perf win here) into https://github.com/radare/radare2/issues/6757 for after flags refactor.

**Comment by radare** [comment]
27 code analysis regressions. we need to check them to see if the test was wrong before or its an issue in this PR. i think master will be green again tonight, so you can rebase after the rest of changes and see if travis situation improves.

**Comment by radare** [comment]
Optionally if you want this PR to be merged now i would suggest to add another config variable under the `anal.` eval space disabled by default. and users can activate that to test it instead of enforcing this change as default.

There are several other anal config vars that should be reviewed and activated/disabled/removed if you are in the mood. As well as improving the new `aab` analysis.

Thanks

**Comment by SchumBlubBlub** [comment]
already looking at the regression.

**Comment by SchumBlubBlub** [comment]
Green now if rebased to before broken in master.

**Pull request merged**